File: sento-runtime/src/main/java/io/sento/annotations/OnTextChanged.java
package io.sento.annotations;

import android.text.TextWatcher;
import android.widget.TextView;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.CLASS)
@Target(ElementType.METHOD)
@ListenerClass(
    owner = TextView.class,
    listener = TextWatcher.class,
    setter = "addTextChangedListener",
    unsetter = "removeTextChangedListener",
    callback = "onTextChanged"
)
public @interface OnTextChanged {
  public int[] value();
}

File: sento-sample/src/main/java/io/sento/sample/annotations/OnTextChanged.kt
package io.sento.sample.annotations

import android.text.TextWatcher
import android.widget.TextView
import io.sento.annotations.ListenerClass

@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
@ListenerClass(
    owner = TextView::class,
    listener = TextWatcher::class,
    setter = "addTextChangedListener",
    unsetter = "removeTextChangedListener",
    callback = "onTextChanged"
)
public annotation class OnTextChanged(vararg val value: Int)

@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
@ListenerClass(
    owner = TextView::class,
    listener = TextWatcher::class,
    setter = "addTextChangedListener",
    unsetter = "removeTextChangedListener",
    callback = "afterTextChanged"
)
public annotation class AfterTextChanged(vararg val value: Int)

@Retention(AnnotationRetention.RUNTIME)
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
@ListenerClass(
    owner = TextView::class,
    listener = TextWatcher::class,
    setter = "addTextChangedListener",
    unsetter = "removeTextChangedListener",
    callback = "beforeTextChanged"
)
public annotation class BeforeTextChanged(vararg val value: Int)
